drm/vc4: Calculate bpc based on max_requested_bpc

HiassofT opened the pull request:
This aligns vc4 with Intel, AMD and Synopsis drivers and fixes max bpc connector property not working as expected as conn_state->max_bpc isn't neccessarily up to date at the point we need the property value.

I initially noticed that on LibreELEC with kernel 6.1 and was wondering why switching to 12bpc didn't work. I've tested this now on RPiOS with 64bit 5.15 kernel on a RPi4 and a 1920x1200 Monitor which supports 12-bit YCC 4:2:2 and it has the same issues.

eg enable drm debugging and set max bpc to 12 - this should switch the output to YCC 422
```
root@raspberrypi:~# echo 0x02 > /sys/module/drm/parameters/debug
root@raspberrypi:~# modetest -M vc4 -w "32:max bpc:12"
```
dmesg shows that vc4 isn't even trying to check for 12-bit support (that's because max_bpc in connector state is stil 8, only max_requested_bpc is 12)
```
[   78.346261] vc4-drm gpu: [drm:vc4_atomic_check [vc4]] crtc-3: Trying to find a channel.
[   78.346376] vc4-drm gpu: [drm:vc4_atomic_check [vc4]] crtc-3: Already enabled, reusing channel 0.
[   78.346504] vc4-drm gpu: [drm:vc4_hdmi_encoder_atomic_check [vc4]] Trying with a 8 bpc output
[   78.346620] vc4-drm gpu: [drm:vc4_hdmi_encoder_atomic_check [vc4]] Trying with an RGB output
[   78.346718] vc4-drm gpu: [drm:vc4_hdmi_sink_supports_format_bpc.isra.29 [vc4]] RGB Format, checking the constraints.
[   78.346790] vc4-drm gpu: [drm:vc4_hdmi_sink_supports_format_bpc.isra.29 [vc4]] RGB format supported in that configuration.
[   78.346862] vc4-drm gpu: [drm:vc4_hdmi_encoder_atomic_check [vc4]] Mode 1920x1200 @ 60Hz: Found configuration: bpc: 8, fmt: RGB, clock: 154000000
```

With this change, using max_requested_bpc, it works as expected:
```
[   53.821214] vc4-drm gpu: [drm:vc4_atomic_check [vc4]] crtc-3: Trying to find a channel.
[   53.821334] vc4-drm gpu: [drm:vc4_atomic_check [vc4]] crtc-3: Already enabled, reusing channel 0.
[   53.821459] vc4-drm gpu: [drm:vc4_hdmi_encoder_atomic_check [vc4]] Trying with a 12 bpc output
[   53.821537] vc4-drm gpu: [drm:vc4_hdmi_encoder_atomic_check [vc4]] Trying with an RGB output
[   53.821634] vc4-drm gpu: [drm:vc4_hdmi_sink_supports_format_bpc.isra.0 [vc4]] RGB Format, checking the constraints.
[   53.821708] vc4-drm gpu: [drm:vc4_hdmi_sink_supports_format_bpc.isra.0 [vc4]] 12 BPC but sink doesn't support Deep Color 36.
[   53.821780] vc4-drm gpu: [drm:vc4_hdmi_encoder_atomic_check [vc4]] Failed, Trying with an YUV422 output
[   53.821870] vc4-drm gpu: [drm:vc4_hdmi_sink_supports_format_bpc.isra.0 [vc4]] YUV422 format, checking the constraints.
[   53.821942] vc4-drm gpu: [drm:vc4_hdmi_sink_supports_format_bpc.isra.0 [vc4]] YUV422 format supported in that configuration.
[   53.822015] vc4-drm gpu: [drm:vc4_hdmi_encoder_atomic_check [vc4]] Mode 1920x1200 @ 60Hz: Found configuration: bpc: 12, fmt: YUV 4:2:2, clock: 154000000
```

The AMD commit 01c22997bed059 has some more info in the commit message and in the comment added to the driver

ping @mripard

HiassofT committed:
drm/vc4: Calculate bpc based on max_requested_bpc

This aligns vc4 with Intel, AMD and Synopsis drivers and fixes max bpc
connector property not working as expected on monitors with YCbCr 4:2:2
support but not deep color support.

max_bpc in connector state is clamped at max_bpc from display info and
the latter only takes deep color modes into account so it will always
be 8, even if the display can do 4:2:2 12-bit output.

Signed-off-by: Matthias Reichl <hias@horus.com>

pelwell commented:
This PR looks very plausible and I'm happy to take it on trust, but does anyone more knowledgeable about DRM have any feedback?

HiassofT commented:
The use case is being able to get YCbCr 4:2:2 on a HDMI monitor with 4:2:2 support but without 10/12 bit deep color support in auto mode when the max_bpc connector property is set to 12.

The max_bpc display info in drm only accounts for deep color modes, so max_bpc will be clamped to 8 which means we will never consider the 12bit 4:2:2 mode in this case.

As we are doing our own 10/12 bit and 4:2:2 support checks and consider 4:2:2 to be a 12-bit mode we can't use the clamped max_bpc.

I fear changing the clamping so that 4:2:2 support on HDMI monitors would set max_bpc in display info might have unwanted side effects on other display drivers as then 10/12 bit 4:4.4 deep color modes might also get selected.

I''m open to rework 4:4:4/4:2:2 selection in auto mode but it's tricky, and I think using max_requested_bpc is the best solution for now.

HiassofT commented:
I've changed the commit message to explain the 4:2:2 / deep color clamping issue and dropped the "not up to date" part.

When looking at the AMD/Intel driver the commits switching to max_requested_bpc mentioned potential problems with non-atomic commits and max_bpc containing old values and initially I thought this is what happened on RPi as well.

But on a second thought and glance through the code I'm no longer sure that's actually the case - the code looks sane to me. I might still be missing something and RPi could also have the issue though.
